### PR TITLE
Extended button added

### PR DESCRIPTION
**Pull request by TomWayne98**

### CODE REVIEW

File: library/src/main/java/com/github/clans/fab/FloatingActionButton.java [hunk 1]
```diff
@@ -106,17 +116,27 @@ public FloatingActionButton(Context context, AttributeSet attrs) {
 
     public FloatingActionButton(Context context, AttributeSet attrs, int defStyleAttr) {
         super(context, attrs, defStyleAttr);
+        if (defStyleAttr == -1){
+            mIsExtended = true;
+        }else {
+            mIsExtended = false;
+        }
+        mContext = context;
         init(context, attrs, defStyleAttr);
```
Comment (TomWayne98) [inline library/src/main/java/com/github/clans/fab/FloatingActionButton.java:138]: I use -1 constant to tell class that it current button will be extended. It called from "createExtendedButton" method in FloatingActionMenu class.

File: library/src/main/java/com/github/clans/fab/FloatingActionMenu.java [hunk 2]
```diff
@@ -234,6 +246,21 @@ public void onAnimationUpdate(ValueAnimator animation) {
...
+        params.height = LayoutParams.WRAP_CONTENT;
+        params.width = LayoutParams.WRAP_CONTENT;
+        setLayoutParams(params);
+        this.startAnimation(AnimationUtils.loadAnimation(getContext(), R.anim.fab_replace_extended));
+    }
+    public void setIconPosition(float yPos,float xPos) {
+        mImageToggle.setX(xPos - mImageToggle.getWidth() / 2 );
+        mImageToggle.setY(yPos - mImageToggle.getHeight() / 2);
+        toggle(true); // Open menu automatically when collapsed
+    }
```
Comment (TomWayne98) [inline library/src/main/java/com/github/clans/fab/FloatingActionMenu.java:261]: Menu is automatically opened when it collapse from its expanded state.

File: library/src/main/java/com/github/clans/fab/FloatingActionMenu.java [hunk 3]
```diff
@@ -125,6 +131,9 @@ public FloatingActionMenu(Context context, AttributeSet attrs, int defStyleAttr)
 
     private void init(Context context, AttributeSet attrs) {
         TypedArray attr = context.obtainStyledAttributes(attrs, R.styleable.FloatingActionMenu, 0, 0);
+        mIsExtended = attr.getBoolean(R.styleable.FloatingActionMenu_menu_isExtended, false);
+        mMenuTextString = attr.getString(R.styleable.FloatingActionMenu_menu_menuText);
+        mMenuTextColor = attr.getInt(R.styleable.FloatingActionMenu_menu_menuTextColor, Color.WHITE);
         mButtonSpacing = attr.getDimensionPixelSize(R.styleable.FloatingActionMenu_menu_buttonSpacing, mButtonSpacing);
```
Comment (TomWayne98) [inline library/src/main/java/com/github/clans/fab/FloatingActionMenu.java:134]: Setters for menu text and default menu shape.

File: library/src/main/java/com/github/clans/fab/FloatingActionMenu.java [hunk 4]
```diff
@@ -696,28 +762,28 @@ public void close(final boolean animate) {
...
+                            @Override
+                            public void run() {
+                                if (!isOpened()) return;
+                                if (fab != mMenuButton) {
+                                    fab.hide(animate);
+                                }
+                                Label label = (Label) fab.getTag(R.id.fab_label);
+                                if (label != null && label.isHandleVisibilityChanges()) {
+                                    label.hide(animate);
+                                }
```
Comment (TomWayne98) [inline library/src/main/java/com/github/clans/fab/FloatingActionMenu.java:783]: Menu should't open in extended state.

File: library/src/main/java/com/github/clans/fab/FloatingActionButton.java [hunk 5]
```diff
@@ -498,9 +554,46 @@ void playShowAnimation() {
...
+                    setMeasuredDimension(calculateMeasuredWidth(), calculateMeasuredHeight());
+                    updateBackground();
+                    ((FloatingActionMenu) getParent()).onExtendedMenuCollapse();
+                    getHandler().postDelayed(new Runnable() {
+                        @Override
+                        public void run() {
+                            ((FloatingActionMenu) getParent()).setIconPosition((getY() + (getHeight()/2)),(getX() + (getWidth()/2))); // There must be some delay to calculate position on final x,y.
+                        }
+                    },300);
+                }
```
Comment (TomWayne98) [inline library/src/main/java/com/github/clans/fab/FloatingActionButton.java:606]: Need to set icon position when animation already ended.

File: library/src/main/java/com/github/clans/fab/CustomShape.java [hunk 6]
```diff
@@ -0,0 +1,31 @@
...
+ *
+ * @author Josef Hruška (josef@stepuplabs.io)
+ */
+
+public class CustomShape extends RectShape {
+    private Context mContext;
+    private float mRadius = 75f;
+    private float mExtraTopShadow = -7f;
+    private float mExtraLeftShadow = -5f;
+    private float mButtonPadding = 64f;
```
Comment (davidvavra) [inline library/src/main/java/com/github/clans/fab/CustomShape.java]: Add paddings as xml dimensions. They can be easily overriden in landscape/tablets etc.

File: library/src/main/java/com/github/clans/fab/FloatingActionButton.java [hunk 7]
```diff
@@ -95,9 +101,14 @@
     private boolean mShouldSetProgress;
     private int mProgressMax = 100;
     private boolean mShowProgressBackground;
+    private boolean mIsExtended = false;
+    private float mLandscapePadding = 60f;
```
Comment (davidvavra) [inline library/src/main/java/com/github/clans/fab/FloatingActionButton.java]: Add paddings as xml dimensions. They can be easily overriden in landscape/tablets etc.

File: library/src/main/java/com/github/clans/fab/FloatingActionButton.java [hunk 8]
```diff
@@ -106,17 +117,27 @@ public FloatingActionButton(Context context, AttributeSet attrs) {
 
     public FloatingActionButton(Context context, AttributeSet attrs, int defStyleAttr) {
         super(context, attrs, defStyleAttr);
+        if (defStyleAttr == -1){
```
Comment (davidvavra) [inline library/src/main/java/com/github/clans/fab/FloatingActionButton.java]: Make the -1 as a named constant

File: library/src/main/java/com/github/clans/fab/FloatingActionButton.java [hunk 9]
```diff
@@ -106,17 +117,27 @@ public FloatingActionButton(Context context, AttributeSet attrs) {
...
     public FloatingActionButton(Context context, AttributeSet attrs, int defStyleAttr, int defStyleRes) {
         super(context, attrs, defStyleAttr, defStyleRes);
+        mIsExtended = false;
         init(context, attrs, defStyleAttr);
     }
 
     private void init(Context context, AttributeSet attrs, int defStyleAttr) {
         TypedArray attr = context.obtainStyledAttributes(attrs, R.styleable.FloatingActionButton, defStyleAttr, 0);
+        if (!mIsExtended){
+        mIsExtended = attr.getBoolean(R.styleable.FloatingActionButton_fab_isExtended, false);
```
Comment (davidvavra) [inline library/src/main/java/com/github/clans/fab/FloatingActionButton.java]: Not formatted

File: library/src/main/java/com/github/clans/fab/FloatingActionButton.java [hunk 10]
```diff
@@ -498,9 +554,46 @@ void playShowAnimation() {
...
+    void playHideExtendedAnimation() {
+        mHideExtendedAnimation.setAnimationListener(new Animation.AnimationListener() {
+            @Override
+            public void onAnimationStart(Animation animation) {
+                //Nothing
+            }
+
+            @Override
+            public void onAnimationEnd(Animation animation) {
+                //TODO add condition
```
Comment (davidvavra) [inline library/src/main/java/com/github/clans/fab/FloatingActionButton.java]: Forgotten todo?

File: library/src/main/java/com/github/clans/fab/FloatingActionButton.java [hunk 11]
```diff
@@ -696,8 +789,13 @@ private void init() {
 
         @Override
         public void draw(Canvas canvas) {
-            canvas.drawCircle(calculateCenterX(), calculateCenterY(), mRadius, mPaint);
-            canvas.drawCircle(calculateCenterX(), calculateCenterY(), mRadius, mErase);
+            if (mIsExtended) {
+                canvas.drawRoundRect(new RectF(0,0,Util.getScreenWidth(getContext()) -  Util.dpToPx(getContext(),mLandscapePadding),Util.dpToPx(getContext(),56f)),70f,70f, mPaint);
+                canvas.drawRoundRect(new RectF(0,0,Util.getScreenWidth(getContext()) -  Util.dpToPx(getContext(),mLandscapePadding),Util.dpToPx(getContext(),56f)),70f,70f, mErase);
```
Comment (davidvavra) [inline library/src/main/java/com/github/clans/fab/FloatingActionButton.java]: Duplicated code, create a method

File: library/src/main/java/com/github/clans/fab/FloatingActionMenu.java [hunk 12]
```diff
@@ -125,6 +130,9 @@ public FloatingActionMenu(Context context, AttributeSet attrs, int defStyleAttr)
 
     private void init(Context context, AttributeSet attrs) {
         TypedArray attr = context.obtainStyledAttributes(attrs, R.styleable.FloatingActionMenu, 0, 0);
+        mIsExtended = attr.getBoolean(R.styleable.FloatingActionMenu_menu_isExtended, false);
+        mMenuTextString = attr.getString(R.styleable.FloatingActionMenu_menu_menuText);
```
Comment (davidvavra) [inline library/src/main/java/com/github/clans/fab/FloatingActionMenu.java]: Better name it "FloatingActionMenu_menu_extendedButtonText"

File: library/src/main/java/com/github/clans/fab/FloatingActionMenu.java [hunk 13]
```diff
@@ -125,6 +130,9 @@ public FloatingActionMenu(Context context, AttributeSet attrs, int defStyleAttr)
...
+        mMenuTextColor = attr.getInt(R.styleable.FloatingActionMenu_menu_menuTextColor, Color.WHITE);
```
Comment (davidvavra) [inline library/src/main/java/com/github/clans/fab/FloatingActionMenu.java]: Rename to FloatingActionMenu_menu_extendedButtonTextColor

**Review by davidvavra — CHANGES_REQUESTED** [review]
Also please edit README.md - add section how to use extended FAB.

File: library/src/main/java/com/github/clans/fab/CustomShape.java (hunk 6 above)
Comment (TomWayne98) [reply library/src/main/java/com/github/clans/fab/CustomShape.java]: I realized that with last changes which I have made it's will use just one padding for both landscape and portrait thus I changed the name to mExtendedButtonPadding.

File: library/src/main/res/values/dimens.xml [hunk 14]
```diff
@@ -5,5 +5,6 @@
     <dimen name="fab_size_mini">40dp</dimen>
     <dimen name="labels_text_size">14sp</dimen>
     <dimen name="extended_button_padding">64dp</dimen>
+    <dimen name="custom_shape_buttonPadding">70dp</dimen>
```
Comment (davidvavra) [inline library/src/main/res/values/dimens.xml]: Better name pls, custom shape would mean that user of the library can customize it. What is it exactly?

File: library/src/main/java/com/github/clans/fab/CustomShape.java [hunk 15]
```diff
@@ -17,15 +17,16 @@
     private float mRadius = 75f;
```
Comment (davidvavra) [inline library/src/main/java/com/github/clans/fab/CustomShape.java]: Also please update README.md and add instructions how to use extended FAB.

**Review by davidvavra — CHANGES_REQUESTED** [review]

File: README.md [hunk 16]
```diff
@@ -109,6 +110,9 @@ Here are all the **FloatingActionMenu**'s xml attributes with their **default va
         android:layout_marginRight="10dp"
         android:layout_marginBottom="10dp"
         android:layout_marginLeft="10dp"
+        fab:menu_isExtended="false"
+        fab:menu_extendedButtonText="Add new expense"
+        fab:menu_extendedButtonTextColor="Color.WHITE"
```
Comment (davidvavra) [inline README.md]: The value is really strange. Adding Java to XML? You mean @color/white?

File: library/src/main/java/com/github/clans/fab/FloatingActionButton.java [hunk 17]
```diff
@@ -106,17 +111,27 @@ public FloatingActionButton(Context context, AttributeSet attrs) {
 
     public FloatingActionButton(Context context, AttributeSet attrs, int defStyleAttr) {
         super(context, attrs, defStyleAttr);
+        if (defStyleAttr == -1){
```
Comment (davidvavra) [inline library/src/main/java/com/github/clans/fab/FloatingActionButton.java]: This is still unnamed constant.

File: library/src/main/java/com/github/clans/fab/FloatingActionButton.java [hunk 18]
```diff
@@ -421,7 +466,12 @@ public void getOutline(View view, Outline outline) {
     }
 
     private Drawable createCircleDrawable(int color) {
-        CircleDrawable shapeDrawable = new CircleDrawable(new OvalShape());
+        CircleDrawable shapeDrawable;
+        if (mIsExtended){
+       shapeDrawable = new CircleDrawable(new ExtendedButtonShadowShape(getContext()));
```
Comment (davidvavra) [inline library/src/main/java/com/github/clans/fab/FloatingActionButton.java]: Wrong formatting

File: library/src/main/java/com/github/clans/fab/FloatingActionButton.java [hunk 19]
```diff
@@ -498,9 +552,45 @@ void playShowAnimation() {
...
+                    FloatingActionButton.super.onAnimationEnd();
+                    setMeasuredDimension(calculateMeasuredWidth(), calculateMeasuredHeight());
+                    updateBackground();
+                    ((FloatingActionMenu) getParent()).onExtendedMenuCollapse();
+                    getHandler().postDelayed(new Runnable() {
+                        @Override
+                        public void run() {
+                            ((FloatingActionMenu) getParent()).setIconPosition((getY() + (getHeight()/2)),(getX() + (getWidth()/2))); // There must be some delay to calculate position on final x,y.
+                        }
+                    },300);
```
Comment (davidvavra) [inline library/src/main/java/com/github/clans/fab/FloatingActionButton.java]: Wrong formatting

File: library/src/main/java/com/github/clans/fab/FloatingActionMenu.java [hunk 20]
```diff
@@ -45,8 +46,10 @@
     private AnimatorSet mCloseAnimatorSet = new AnimatorSet();
     private AnimatorSet mIconToggleSet;
 
+    private boolean mIsExtended = false;
+    private String extendedButtonText = "Add new expense";
```
Comment (davidvavra) [inline library/src/main/java/com/github/clans/fab/FloatingActionMenu.java]: Hardcoding our text into the library? Replace with null value please.

File: library/src/main/java/com/github/clans/fab/FloatingActionMenu.java [hunk 21]
```diff
@@ -259,10 +285,36 @@ private void createMenuButton() {
...
+            mMenuButton.mShadowYOffset = Util.dpToPx(getContext(), mMenuShadowYOffset);
+        }
+        mMenuButton.setColors(mMenuColorNormal, mMenuColorPressed, mMenuColorRipple);
+        mMenuButton.mShadowColor = mMenuShadowColor;
+        mMenuButton.updateBackground();
+        mMenuButton.setLabelText(mMenuLabelText);
+
+        mMenuText = new TextView(getContext());
+        mMenuText.setText(extendedButtonText);
+        mMenuText.setTextSize(16);
```
Comment (davidvavra) [inline library/src/main/java/com/github/clans/fab/FloatingActionMenu.java]: Make text size configurable as well, it's better to specify it in xml.

File: library/src/main/java/com/github/clans/fab/FloatingActionMenu.java [hunk 22]
```diff
@@ -447,28 +509,26 @@ protected void onFinishInflate() {
...
-            addLabel(fab);
-
-            if (fab == mMenuButton) {
-                mMenuButton.setOnClickListener(new OnClickListener() {
-                    @Override
-                    public void onClick(View v) {
-                        toggle(mIsAnimated);
-                    }
-                });
+            if(getChildAt(i) instanceof FloatingActionButton) {
```
Comment (davidvavra) [inline library/src/main/java/com/github/clans/fab/FloatingActionMenu.java]: Wrong formatting

File: library/src/main/java/com/github/clans/fab/FloatingActionMenu.java [hunk 23]
```diff
@@ -616,11 +674,15 @@ public boolean isOpened() {
 
     public void toggle(boolean animate) {
         if (isOpened()) {
-            close(animate);
+                close(animate);
         } else {
             open(animate);
+            if (mIsExtended) {
```
Comment (davidvavra) [inline library/src/main/java/com/github/clans/fab/FloatingActionMenu.java:682]: Wrong formatting

File: library/src/main/java/com/github/clans/fab/FloatingActionMenu.java [hunk 24]
```diff
@@ -640,6 +702,8 @@ public void open(final boolean animate) {
             int delay = 0;
             int counter = 0;
             mIsMenuOpening = true;
+
+            if (!mIsExtended) {
```
Comment (davidvavra) [inline library/src/main/java/com/github/clans/fab/FloatingActionMenu.java:708]: wrong formatting

File: library/src/main/java/com/github/clans/fab/FloatingActionMenu.java [hunk 25]
```diff
@@ -970,6 +1035,14 @@ public void removeMenuButton(FloatingActionButton fab) {
         mButtonsCount--;
     }
 
+    public void setMenuText(String menuText) {
```
Comment (davidvavra) [inline library/src/main/java/com/github/clans/fab/FloatingActionMenu.java]: Rename to setExtendedButtonText

File: library/src/main/java/com/github/clans/fab/FloatingActionMenu.java [hunk 26]
```diff
@@ -970,6 +1035,14 @@ public void removeMenuButton(FloatingActionButton fab) {
...
+        extendedButtonText = menuText;
+    }
+
+    public void setMenuTextColor(int menuTextColor) {
```
Comment (davidvavra) [inline library/src/main/java/com/github/clans/fab/FloatingActionMenu.java]: Rename to setExtendedButtonTextColor

File: library/src/main/res/anim/fab_replace_extended.xml [hunk 27]
```diff
@@ -0,0 +1,9 @@
+<?xml version="1.0" encoding="utf-8"?>
```
Comment (davidvavra) [inline library/src/main/res/anim/fab_replace_extended.xml]: Rename to fab_extended_replace

File: library/src/main/res/values/dimens.xml [hunk 28]
```diff
@@ -4,5 +4,7 @@
     <dimen name="fab_size_normal">56dp</dimen>
     <dimen name="fab_size_mini">40dp</dimen>
     <dimen name="labels_text_size">14sp</dimen>
+    <dimen name="extended_button_padding">64dp</dimen>
+    <dimen name="extend_button_shadow_padding">70dp</dimen>
```
Comment (davidvavra) [inline library/src/main/res/values/dimens.xml]: Rename to extended_button_shadow_padding

File: sample/src/main/res/anim/fab_replace_extended.xml [hunk 29]
```diff
@@ -0,0 +1,7 @@
+<?xml version="1.0" encoding="utf-8"?>
```
Comment (davidvavra) [inline sample/src/main/res/anim/fab_replace_extended.xml]: Why do we need this in the sample? It should be in the library imho.

**Review by davidvavra — CHANGES_REQUESTED** [review]

**Pull request merged**